### PR TITLE
Create label and undistort

### PR DESCRIPTION
**Pull request by ravern**
Creates the `label` and `undistort` subcommands.

A typical workflow will contain the following:
1. Put all the images into one directory
2. Run the `label` tool, which will open up a window
3. Click the corners from **left to right** and **top to bottom** (Press "backspace" to undo)
4. Pass the resulting CSV file into `undistort`

The available options for `label` are as follows:
- `-p` - The directory of images to label (e.g. `images`)
- `-f` - The format of the images (e.g. `jpg`, `png`)
- `-o` - The output file (e.g. `labels.csv`)
- `-d` - The dimensions of the board (e.g. `8x8`)

The available options for `undistort` are as follows:
- `-p` - The directory of images to label (e.g. `images`)
- `-f` - The format of the images (e.g. `jpg`, `png`)
- `-l` - The file containing the labels (e.g. `labels.csv`)
- `-d` - The dimensions of the board (e.g. `8x8`)
- `-r` - Whether to crop or letterbox the undistorted images

_"Crop" means that the extra pixels (due to the curves) will be discarded, while "Letterbox" means that black pixels will be inserted._

### CODE REVIEW

File: cam_tools/undistort.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    # Calculate the camera calibration.
+    ret, mtx, dist, rvecs, tvecs = cv2.calibrateCamera(
+        object_points, image_points, image.shape[::-1][1:3], None, None)
+    if not ret:
+        raise CamToolsError("failed to caculate the camera calibration")
+
+    # Turn `retain pixels` into an integer.
+    if retain_pixels:
+        retain_pixels = 1
```
Comment (mindworker) [inline cam_tools/undistort.py]: Better use a new variable than overwriting with a different type 😅
Something like

```is_retain = 1 if retain_pixels else 0```

File: cam_tools/utils.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+from cam_tools.error import CamToolsError
+
+
+def parse_dimension(dimension):
```
[the remark this answers is not in the document]
Comment (ravernkoh) [reply cam_tools/utils.py]: I'm not 100% sure how to write tests in Python. What is the norm?

File: cam_tools/label.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        cv2.imshow(WINDOW_NAME, images[len(images) - 1])
+        key = cv2.waitKey(0)
+
+        # If "backspace" was pressed, remove one image (and corner).
+        if key == 8 and len(corners) > 0:
+            corners.pop()
+            images.pop()
+
+        # Check whether there are enough corners.
+        if len(corners) == dimension[0] * dimension[1]:
```
[the remark this answers is not in the document]
Comment (ravernkoh) [reply cam_tools/label.py]: It's worth noting that the code is actually different if this is moved into the `while` condition. However, in this case it doesn't matter.

File: cam_tools/label.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    images = [image]
+
+    # Bind the mouse callback to the current state (i.e. corners and images).
+    cv2.setMouseCallback(WINDOW_NAME, lambda event, x, y, flags, param: handle_click(
+        corners, images, event, x, y, dimension))
+
+    # Display the image and collect clicks.
+    while True:
+        # Display the latest window.
+        cv2.imshow(WINDOW_NAME, images[len(images) - 1])
```
[the remark this answers is not in the document]
Comment (ravernkoh) [reply cam_tools/label.py]: Completely forgot this existed. The wonders of Python 🤤

**Pull request merged**